File: benchmarks/go/fannkuch.go
// go build -o fannkuch_go fannkuch.go
// Canonical fannkuch-redux (Benchmarks Game shape), single-threaded.
package main

import "fmt"

func main() {
	const n = 10
	perm1 := make([]int64, n)
	count := make([]int64, n)
	perm := make([]int64, n)
	for i := 0; i < n; i++ {
		perm1[i] = int64(i)
	}
	var maxFlips, checksum, permCount int64
	r := n
	for {
		for r != 1 {
			count[r-1] = int64(r)
			r--
		}
		copy(perm, perm1)
		var flips int64
		k := perm[0]
		for k != 0 {
			a, b := 0, int(k)
			for a < b {
				perm[a], perm[b] = perm[b], perm[a]
				a++
				b--
			}
			flips++
			k = perm[0]
		}
		if flips > maxFlips {
			maxFlips = flips
		}
		if permCount%2 == 0 {
			checksum += flips
		} else {
			checksum -= flips
		}
		for {
			if r == n {
				fmt.Println(checksum)
				fmt.Printf("Pfannkuchen(%d) = %d\n", n, maxFlips)
				return
			}
			perm0 := perm1[0]
			for i := 0; i < r; i++ {
				perm1[i] = perm1[i+1]
			}
			perm1[r] = perm0
			count[r]--
			if count[r] > 0 {
				break
			}
			r++
		}
		permCount++
	}
}

File: benchmarks/rust/fannkuch.rs
// rustc -O fannkuch.rs -o fannkuch_rust
// Canonical fannkuch-redux (Benchmarks Game shape), single-threaded.
fn main() {
    let n: usize = 10;
    let mut perm1: Vec<i64> = (0..n as i64).collect();
    let mut count = vec![0i64; n];
    let mut perm = vec![0i64; n];
    let mut max_flips: i64 = 0;
    let mut checksum: i64 = 0;
    let mut perm_count: i64 = 0;
    let mut r = n;
    loop {
        while r != 1 {
            count[r - 1] = r as i64;
            r -= 1;
        }
        perm.copy_from_slice(&perm1);
        let mut flips: i64 = 0;
        let mut k = perm[0];
        while k != 0 {
            perm[..(k as usize + 1)].reverse();
            flips += 1;
            k = perm[0];
        }
        if flips > max_flips {
            max_flips = flips;
        }
        checksum += if perm_count % 2 == 0 { flips } else { -flips };
        loop {
            if r == n {
                println!("{}", checksum);
                println!("Pfannkuchen({}) = {}", n, max_flips);
                return;
            }
            let perm0 = perm1[0];
            for i in 0..r {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = perm0;
            count[r] -= 1;
            if count[r] > 0 {
                break;
            }
            r += 1;
        }
        perm_count += 1;
    }
}
